src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -182,48 +184,77 @@ private void ThrowIfDisposed()
             }
         }
 
-        private static void Replenish(DefaultPartitionedRateLimiter<TResource, TKey> limiter)
+        private async Task TimerLoop()
         {
-            lock (limiter.Lock)
+            lock (Lock)
             {
-                if (limiter._disposed)
+                if (_disposed)
                 {
                     return;
                 }
 
                 // If the cache has been invalidated we need to recreate it
-                if (limiter._cacheInvalid)
+                if (_cacheInvalid)
                 {
-                    limiter._cachedLimiters.Clear();
-                    bool cacheStillInvalid = false;
-                    foreach (KeyValuePair<TKey, Lazy<RateLimiter>> kvp in limiter._limiters)
+                    _cachedLimiters.Clear();
+                    _cachedLimiters = _limiters.ToList();
+                }
+            }
+
+            List<RateLimiter>? limitersToDispose = null;
+
+            // cachedLimiters is safe to use outside the lock because it is only updated by the Timer
+            foreach (KeyValuePair<TKey, Lazy<RateLimiter>> rateLimiter in _cachedLimiters)
```

Comment:
```suggestion
            foreach ((TKey key, Lazy<RateLimiter> lazyLimiter) in _cachedLimiters)
```

Or does our netstandardness mean KVP doesn't have a `Deconstruct` method?

Reply:
Correct, adding in ns2.1
https://apisof.net/catalog/e3b857aa-bf36-597f-b93f-df8f95512c93?fx=netstandard2.1